### PR TITLE
project

### PR DESCRIPTION
**Pull request by aric87**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+env 
+env
```
Comment (ripleyaffect) [inline .gitignore]: duped

File: blog/__init__.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import os
+from flask import Flask
```
Comment (ripleyaffect) [inline blog/__init__.py]: imports are placed in three groups, with one space between each group. the first is standard library imports. (i.e. `os`), the second is third party imports (i.e. `flask`), the third is project imports (i.e. if you're importing `blog` elsewhere in this app). There are two spaces after the imports sections. Each module should also start with a docstring at the very top

File: blog/__init__.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+app = Flask(__name__)
+config_path = os.environ.get("CONFIG_PATH","blog.config.DevelopmentConfig")
```
Comment (ripleyaffect) [inline blog/__init__.py]: spaces after params (i.e. after `"CONFIG_PATH"`)

File: blog/__init__.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+app.config.from_object(config_path)
+
+from . import views
+from . import filters
```
Comment (ripleyaffect) [inline blog/__init__.py]: this is the same as saying `import views`/ `import filters` and doesn't look like it's being used here?

File: blog/__init__.py [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+
```
Comment (ripleyaffect) [inline blog/__init__.py]: There should always be one empty line at the end of a file

File: blog/database.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+from . import app
+
+engine = create_engine(app.config["SQLALCHEMY_DATABASE_URI"])
+Base = declarative_base()
+Session = sessionmaker(bind=engine)
+session = Session()
+
+import datetime
+from sqlalchemy import Column, Integer, String, Text, DateTime
```
Comment (ripleyaffect) [inline blog/database.py]: all imports should be at the top

File: blog/config.py [hunk 7]
```diff
@@ -0,0 +1,4 @@
+import os
+class DevelopmentConfig(object):
+    SQLALCHEMY_DATABASE_URI = "postgresql://ubuntu:thinkful@localhost:5432/blogful"
+    EBUG = True
```
Comment (ripleyaffect) [inline blog/config.py]: `DEBUG`*

File: blog/database.py [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+class Entry(Base):
+    __tablename__= "entries"
+    
+    id =Column(Integer,primary_key = True)
```
Comment (ripleyaffect) [inline blog/database.py]: spaces after the assignment op (`+`)

File: blog/templates/base.html [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+ 
+        <!-- CSS -->
+        <!-- Bootstrap -->
+        <link rel="stylesheet" href="//netdna.bootstrapcdn.com/bootstrap/3.1.1/css/bootstrap.min.css">
+        <!-- Blog -->
+        <link rel="stylesheet" href="{{ url_for('static', filename='css/main.css') }}">
+
+        <!-- JavaScript -->
+        <!-- jQuery -->
+        <script src="http://code.jquery.com/jquery-2.1.1.min.js"></script>
```
Comment (ripleyaffect) [inline blog/templates/base.html]: using a cdn will probably be faster

File: blog/templates/entries.html [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+{% extends "base.html" %}
+{% block content %}
+
+{% for entry in entries %}
+{{ macros.render_entry(entry) }}
+{% endfor %}
+ 
+
+
+
```
Comment (ripleyaffect) [inline blog/templates/entries.html]: just need 1 or 2 spaces

File: blog/views.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
+from flask import render_template, request, redirect, url_for
+from . import app
+from .database import session, Entry
+
+PAGINATE_BY = 20
+def get_entry(id):
```
Comment (ripleyaffect) [inline blog/views.py]: two lines between things on the root level, one otherwise

File: blog/views.py [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+    return session.query(Entry).filter_by(id=id).first()
```
Comment (ripleyaffect) [inline blog/views.py]: This is equivalent to doing `Entry.query.get(id)`

File: blog/views.py [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+
+    total_pages = (count - 1) / limit + 1
+    has_next = page_index < total_pages - 1
+    has_prev = page_index > 0
+
+    entries = session.query(Entry)
+    entries = entries.order_by(Entry.datetime.desc())
+    entries = entries[start:end]
+    if limit == PAGINATE_BY:
+        limit = None
```
Comment (ripleyaffect) [inline blog/views.py]: flask-SQLAlchemy provides a [`paginate`](http://flask-sqlalchemy.pocoo.org/2.1/api/#flask.ext.sqlalchemy.BaseQuery.paginate) function for things like this

File: blog/views.py [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+    entry = Entry(title=request.form["title"],
+        content=request.form["content"]
+    )
+    session.add(entry)
+    session.commit()
+    return redirect(url_for("entries"))
+
+@app.route('/entry/<int:id>')
+def single_entry_get(id):
+    entry=get_entry(id)
```
Comment (ripleyaffect) [inline blog/views.py]: spaces around the assignment op

File: blog/views.py [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+    print(entry)
```
Comment (ripleyaffect) [inline blog/views.py]: print alert!

File: blog/views.py [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+    return render_template(
+        "single_entry.html",
+        entry=entry
+        )
```
Comment (ripleyaffect) [inline blog/views.py]: should be up on the previous line

File: blog/views.py [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+    entry = get_entry(id)
+    return render_template("edit_entry.html",entry=entry)
+    
+@app.route('/entry/<int:id>/edit',methods=['POST'])
+def update_entry(id):
+    entry = get_entry(id)
+    entry.title = request.form["title"]
+    entry.content =request.form["content"]
+    
+    print(entry)
```
Comment (ripleyaffect) [inline blog/views.py]: print alert!

File: blog/views.py [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+        )
+@app.route('/entry/<int:id>/edit',methods=['GET'])
+def edit_entry(id):
+    entry = get_entry(id)
+    return render_template("edit_entry.html",entry=entry)
+    
+@app.route('/entry/<int:id>/edit',methods=['POST'])
+def update_entry(id):
+    entry = get_entry(id)
+    entry.title = request.form["title"]
```
Comment (ripleyaffect) [inline blog/views.py]: should check for existence of `entry` and return a 404 if it doesn't exist

File: blog/views.py [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+from .database import session, Entry
+
+PAGINATE_BY = 20
+def get_entry(id):
+    return session.query(Entry).filter_by(id=id).first()
+    
+    
+@app.route("/")
+@app.route("/page/<int:page>")
+def entries(page=1):
```
Comment (ripleyaffect) [inline blog/views.py]: all functions should have docstrings (triple quoted strings on first line of function)

File: blog/__init__.py (hunk 4 above)
Comment (aric87) [inline blog/__init__.py]: That's the way the Python course teaches it. Not sure why it's there.

File: manage.py [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+
+@manager.command
+def run():
+    port = int(os.environ.get('PORT',8080))
+    app.run(host='0.0.0.0',port=port)
+    
+@manager.command
+def seed():
+    content = """Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."""
+    for i in range(25):
```
Comment (ripleyaffect) [inline manage.py]: passing in an int to `seed` will let you customize the number of posts created. Also using `xrange` here will save you space (doesn't create list, just uses a generator)

File: manage.py [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+from flask.ext.script import Manager
+from blog import app
+from blog.database import session,Entry
+
+manager = Manager(app)
+
+@manager.command
+def run():
+    port = int(os.environ.get('PORT',8080))
+    app.run(host='0.0.0.0',port=port)
```
Comment (ripleyaffect) [inline manage.py]: this IP should be in a constant at the top of the file

File: manage.py [hunk 22]
```diff
@@ -0,0 +1,25 @@
+import os
+from flask.ext.script import Manager
+from blog import app
+from blog.database import session,Entry
+
+manager = Manager(app)
+
+@manager.command
+def run():
+    port = int(os.environ.get('PORT',8080))
```
Comment (ripleyaffect) [inline manage.py]: this default port should be in a constant at the top of the file

**Comment by ripleyaffect** [comment]
Mostly style comments. Logic looks pretty solid :+1:

File: blog/views.py (hunk 13 above)
Comment (aric87) [inline blog/views.py]: that was straight from the curriculum :)

File: blog/views.py (hunk 12 above)
Comment (aric87) [inline blog/views.py]: so, I tried that and it said type object 'Entry' has no attribute 'query'

Comment (ripleyaffect) [inline blog/views.py]: oh, because the model is inheriting from `Base` instead of `Model`, ok

**Pull request merged**